components/script/script_module.rs

```diff
@@ -0,0 +1,1313 @@
+                    AlreadyInCompartment::assert(&*global);
+                    let _ais = AutoIncumbentScript::new(&*global);
+
+                    let abv = CreateAutoObjectVector(*global.get_cx());
+
+                    for descendant in descendants {
+                        assert!(AppendToAutoObjectVector(
+                            abv as *mut AutoObjectVector,
+                            descendant.promise_obj().get()
+                        ));
+                    }
+
+                    rooted!(in(*global.get_cx()) let raw_promise_all = GetWaitForAllPromise(*global.get_cx(), abv));
+
+                    let promise_all =
+                        Promise::new_with_js_promise(raw_promise_all.handle(), global.get_cx());
+
+                    let promise = module_tree.get_promise().borrow();
+                    let promise = promise.as_ref().unwrap().clone();
+
+                    let resolve_promise = TrustedPromise::new(promise.clone());
+                    let reject_promise = TrustedPromise::new(promise.clone());
+
+                    let handler = PromiseNativeHandler::new(
+                        &global,
+                        Some(ModuleHandler::new(Box::new(
+                            task!(all_fetched_resolve: move || {
+                                println!("promise all fetched");
+                                let promise = resolve_promise.root();
+                                promise.resolve_native(&());
+                            }),
+                        ))),
+                        Some(ModuleHandler::new(Box::new(
+                            task!(all_failure_reject: move || {
+                                println!("promise all failed");
+                                let promise = reject_promise.root();
+                                promise.reject_native(&());
+                            }),
+                        ))),
+                    );
+
+                    promise_all.append_native_handler(&handler);
+
+                    return Some(promise_all);
+                }
+            }
+        },
+        Err(err) => {
+            module_tree.set_error(Some(err));
+        },
+    }
+
+    None
+}
+
+#[allow(unsafe_code)]
+/// https://html.spec.whatwg.org/multipage/#fetch-the-descendants-of-a-module-script
+fn fetch_module_descendants(
+    owner: &ModuleOwner,
+    module_tree: &ModuleTree,
+    destination: Destination,
+    mut visited: HashSet<ServoUrl>,
+) -> Result<Vec<Rc<Promise>>, ModuleError> {
+    println!("Start to load dependencies of {}", module_tree.url.clone());
+
+    let global = owner.global();
+
+    module_tree.set_status(ModuleStatus::FetchingDescendants);
+
+    module_tree
+        .resolve_requested_modules(&global, &mut visited)
+        .map(|requested_urls| {
+            module_tree.append_descendant_urls(requested_urls.clone());
+
+            requested_urls
+                .iter()
+                .map(|requested_url| {
```

Comment:
this probably could be a separate function?

Reply:
I just realized I should handle those promise checking in `fetch_single_module_script` and that's how `Step 2` and `Step 3` of [fetch-a-single-module-script](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script) works! Fix at f882ae7. Does that change make sense to you 👀?